src/demo.rs

```diff
@@ -22,45 +21,98 @@ where
 {
     electrum_client: RefCell<Client<T>>,
+    utxo_set: RefCell<Vec<OutPoint>>,
+    capacity: usize,
 }
 
+const DEFAULT_CAPACITY: usize = 10;
+
 impl<T> ElectrumBlockchain<T>
 where
     T: Read + Write,
 {
     pub fn new(electrum_client: Client<T>) -> Self {
-        ElectrumBlockchain { electrum_client: RefCell::new(electrum_client) }
+        Self::with_capacity(electrum_client, DEFAULT_CAPACITY)
+    }
+
+    pub fn with_capacity(electrum_client: Client<T>, capacity: usize) -> Self {
+        ElectrumBlockchain {
+            electrum_client: RefCell::new(electrum_client),
+            utxo_set: RefCell::new(Vec::with_capacity(capacity)),
+            capacity,
+        }
     }
 }
 
 impl<T> Blockchain for ElectrumBlockchain<T>
 where
     T: Read + Write,
 {
-    type Error = (); //TODO: maybe use electrum_client::types::Error ?
+    type Error = super::Error;
 
     fn get_tx(&self, txid: &Txid) -> Result<Transaction, Self::Error> {
-        self.electrum_client.borrow_mut().transaction_get(txid).map_err(|_| ())
+        self.electrum_client.borrow_mut().transaction_get(txid).map_err(|x| x.into())
     }
 
     fn is_unspent(&self, txout: &OutPoint) -> Result<bool, Self::Error> {
         let script = &self.get_tx(&txout.txid)?.output[txout.vout as usize].script_pubkey;
-        let unspent_utxos = &self.electrum_client.borrow_mut().script_list_unspent(&script).map_err(|_| ())?;
-        Ok(unspent_utxos.into_iter().filter(|x| x.tx_hash == txout.txid).count() > 0)
+        let unspent_utxos = &self
+            .electrum_client
+            .borrow_mut()
+            .script_list_unspent(&script)?;
+        Ok(unspent_utxos
+            .into_iter()
+            .filter(|x| x.tx_hash == txout.txid)
+            .count()
+            > 0)
     }
 
-    fn get_random_utxo(&self) -> Result<OutPoint, Self::Error> {
-        Ok(OutPoint {
-            txid: Txid::from_hex(
-                "0f3fb1116e30963f1dc6631ad0cd7f00e324de7f3348264a1bba539fb4721c5d",
-            )
-            .unwrap(),
-            vout: 0,
-        })
+    fn get_random_utxo(&self, seed: &OutPoint) -> Result<Option<OutPoint>, Self::Error> {
+        if self.utxo_set.borrow().len() == 0 {
+            let mut txid = &seed.txid;
+            let mut tx = self.get_tx(&txid)?.clone();
+            let mut scripts = HashSet::new();
+            let coinbase_txid =
+                Txid::from_hex("0000000000000000000000000000000000000000000000000000000000000000")
+                    .unwrap();
+            while self.utxo_set.borrow().len() < self.capacity {
+                txid = &tx.input[0].previous_output.txid;
+
+                // We hit a coinbase!
+                if txid == &coinbase_txid {
+                    break;
+                }
+
+                tx = self.get_tx(&txid)?.clone();
+                scripts.extend(
+                    tx.output
+                        .iter()
+                        .map(|x| x.script_pubkey.clone())
```

Comment:
Nope: since I'm using `tx.is_coin_base` (instead of `txid == coinbase`, as you suggested), tx is used at the next iteration of the loop, thus it can't be partially moved using `into_iter`